File: app/server/api/models/flightModel.js
// Created by Gal Sternad, 26/04/2023
const mongoose = require('mongoose');
const Schema   = mongoose.Schema;

const flightSchema = new Schema({
    // ID leta, ki ga pridobi API
    flightId: {
        type: String,
        required: true
    },
    status: {
        // pove, če je letalo aktivno
        live: {
            type: Boolean,
            required: true
        }
        // TODO: dodaj ostale smiselne lastnosti
    },
    aircraft: {
        type: Schema.Types.ObjectId,
        ref: 'Aircraft',
        required: true
    },
    airport: {
        origin: {
            type: Schema.Types.ObjectId,
            ref: 'Airport',
            required: true
        },
        destination: {
            type: Schema.Types.ObjectId,
            ref: 'Airport',
            required: true
        }
    },
    time: {
        scheduled: {
            departure: {
                type: Date,
                required: true
            },
            arrival: {
                type: Date,
                required: true
            }
        },
        real: {
            departure: {
                type: Date,
                required: true
            },
            arrival: {
                type: Date,
                required: true
            }
        },
        estimated: {
            departure: {
                type: Date,
                required: true
            },
            arrival: {
                type: Date,
                required: true
            }
        },
        metadata: {
            updated: {
                type: Date,
                required: true
            }
        }
    },
    trail: [
        {
            latitude: {
                type: Number,
                required: true
            },
            longitude: {
                type: Number,
                required: true
            },
            altitude: {
                type: Number,
                required: true
            },
            speed: {
                type: Number,
                required: true
            },
            timestamp: {
                type: Date,
                required: true
            },
            // smer letala v stopinjah
            direction: {
                type: Number,
                required: true
            }
        }
    ],
    firstTimestamp: {
        type: Date,
        required: true
    }
}, {timestamps : true});

module.exports = mongoose.model('Flight', flightSchema);

File: api/rest/models/flight_model.py
from mongoengine import *


# ID data of the flight
class FlightIdentification(EmbeddedDocument):
    flight_id = StringField(required=True)


# status of the flight
class FlightStatus(EmbeddedDocument):
    live = BooleanField(required=True)


# airport metadata
class FlightAirportInfo(EmbeddedDocument):
    terminal = StringField(required=True)
    baggage = StringField(required=True)
    gate = StringField(required=True)


# data about flight's origin and destination airports
class FlightAirport(EmbeddedDocument):
    airport = ObjectIdField(required=True)
    info = EmbeddedDocument(FlightAirportInfo)


#
class FlightTime(EmbeddedDocument):
    departure = IntField(required=True)
    arrival = IntField(required=True)


#
class FlightTimeData(EmbeddedDocument):
    scheduled = EmbeddedDocumentField(FlightTime)
    real = EmbeddedDocumentField(FlightTime)
    estimated = EmbeddedDocumentField(FlightTime)


# aircraft's position
class FlightTrail(EmbeddedDocument):
    latitude = FloatField(required=True)
    longitude = FloatField(required=True)
    altitude = IntField(required=True)
    speed = IntField(required=True)
    timestamp = IntField(required=True)
    # current direction of the airplane (in degrees)
    heading = IntField(required=True)


# main flight schema
class Flight(Document):
    identification = EmbeddedDocumentField(FlightIdentification)
    status = EmbeddedDocumentField(FlightStatus)
    aircraft = ObjectIdField(required=True)
    owner = StringField(required=True)
    airspace = StringField(required=True)
    origin = EmbeddedDocumentField(FlightAirport)
    destination = EmbeddedDocumentField(FlightAirport)
    time = EmbeddedDocumentField(FlightTime)
    trail = EmbeddedDocumentListField(FlightTrail)
    first_timestamp = IntField(required=True)
